### PR TITLE
[IMP] point_of_sale,pos_restaurant: cancelled order to preparation tools

### PR DESCRIPTION
**Pull request by davidmonnom**
Before when you deleted a order in PoS with the preparation tools enabled. The order was not cancelled on them.
Now, when cancelling/deleting an order in PoS the information is sent to the preparation tools.

taskId: 3278081

Linked PR: https://github.com/odoo/enterprise/pull/40673

### CODE REVIEW

**Comment by caburj** [comment]
Is there no way to write test for this dev?

**Comment by davidmonnom** [comment]
> Is there no way to write test for this dev?

Done and green :)

File: addons/point_of_sale/static/src/js/Screens/TicketScreen/TicketScreen.js [hunk 1]
```diff
@@ -148,6 +148,9 @@ export class TicketScreen extends IndependentToOrderScreen {
         }
         const { globalState } = this.pos;
         if (order && (await this._onBeforeDeleteOrder(order))) {
+            if (Object.keys(order.lastOrderPrepaChange)) {
```
Comment (caburj) [inline addons/point_of_sale/static/src/js/Screens/TicketScreen/TicketScreen.js:151]: Do you intend to compare the length to zero like so: `Object.keys(order.lastOrderPrepaChange).length !== 0)`?

**Comment by caburj** [comment]
> > Is there no way to write test for this dev?
> 
> Done and green :) [odoo/enterprise#40673](https://github.com/odoo/enterprise/pull/40673)

On it.

**Review by caburj — CHANGES_REQUESTED** [review]
If `pos_preparation_display` is uninstalled, the feature the cancels the order from the kitchen printer doesn't work. It thinks that there are no changes to print. Here are the steps to reproduce:

1. Install `pos_restaurant`.
2.  Uninstall `pos_preparation_display`.
3. Setup a preparation printer.
4. Open a bar.
5. Add a product that can be sent to the preparation printer.
6. Order. (Error popup but it's okay, the order is sent.)
7. [ISSUE] When deleting the order, there is no "changes to print" that is sent to the printer.

**Comment by davidmonnom** [comment]
> If `pos_preparation_display` is uninstalled, the feature that cancels the order from the kitchen printer doesn't work. It thinks that there are no changes to print. Here are the steps to reproduce:
> 
> 1. Install `pos_restaurant`.
> 2. Uninstall `pos_preparation_display`.
> 3. Setup a preparation printer.
> 4. Open a bar.
> 5. Add a product that can be sent to the preparation printer.
> 6. Order. (Error popup but it's okay, the order is sent.)
> 7. [ISSUE] When deleting the order, there is no "changes to print" that is sent to the printer.

Should be good

**Comment by caburj** [comment]
> Should be good

Yeah, maybe it's already printing. But I there is something wrong in the items being printed.

I tested this workflow for 2 situations - with and without pos_preparation_display.
### Steps
1. Create an order, add 1 item and send for preparation (click Order button).
2. Delete the order and check the result of `changesToOrder` which is used for both the preparation printers and displays.

### Results

*With pos_preparation_display*

Screenshot 2023-05-22 at 16 46 34

*Without pos_preparation_display*

Screenshot 2023-05-22 at 16 44 01

### Conclusion

I think regardless of presence of `pos_preparation_display`, the result of `changesToOrder` when deleting the order should be the same. In this case, they differ in sign. Basically, the kitchen printer should print the same thing for the same workflow.

You might be doing something wrong in the calculation of `changesToPrint` when considering the `cancelled` param.

File: addons/point_of_sale/static/src/js/models.js [hunk 2]
```diff
@@ -3142,7 +3144,7 @@ export class Order extends PosModel {
             if (lineChange["quantity"] > 0 && !cancelled) {
                 toAdd.push(lineChange);
             } else {
-                lineChange["quantity"] *= -1; // we change the sign because that's how it is
+                lineChange["quantity"] = Math.abs(lineChange["quantity"]); // we need always positive values.
```
Comment (caburj) [inline addons/point_of_sale/static/src/js/models.js]: Can you quickly explain me why it's sometimes negative? I feel like we are just masking the problem by taking the absolute value.

**Review by caburj — APPROVED** [review]
@robodoo r+

**Pull request closed without merging**